test/webauthn/apis.test.js

```diff
@@ -46,104 +42,730 @@ describe(`apisFunctions: ${printPath("[test/webauthn/apis.test.js]")}`, function
+        });
+
+        it("test signIn fail with wrong credential", async function () {
+            const connectionURI = await startST();
+
+            const origin = "https://supertokens.io";
+            const rpId = "supertokens.io";
+            const rpName = "SuperTokensplm";
+
+            STExpress.init({
+                supertokens: {
+                    connectionURI,
+                },
+                appInfo: {
+                    apiDomain: "api.supertokens.io",
+                    appName: "SuperTokensplm",
+                    websiteDomain: "supertokens.io",
+                },
+                recipeList: [
+                    Session.init(),
+                    WebAuthn.init({
+                        getOrigin: async () => {
+                            return origin;
+                        },
+                        getRelyingPartyId: async () => {
+                            return rpId;
+                        },
+                        getRelyingPartyName: async () => {
+                            return rpName;
+                        },
+                    }),
+                ],
+            });
+
+            // run test if current CDI version >= 2.11
+            if (!(await isCDIVersionCompatible("2.11"))) return;
+
+            const app = express();
+            app.use(middleware());
+            app.use(errorHandler());
+
+            const email = `${Math.random().toString().slice(2)}@supertokens.com`;
+            let registerOptionsResponse = await new Promise((resolve) =>
+                request(app)
+                    .post("/auth/webauthn/options/register")
+                    .send({
+                        email,
+                    })
+                    .expect(200)
+                    .end((err, res) => {
+                        if (err) {
+                            console.log(err);
+                            resolve(undefined);
+                        } else {
+                            resolve(JSON.parse(res.text));
+                        }
+                    })
+            );
+            assert(registerOptionsResponse.status === "OK");
+
+            let signInOptionsResponse = await new Promise((resolve) =>
+                request(app)
+                    .post("/auth/webauthn/options/signin")
+                    .send({ email: email + "wrong" })
+                    .expect(200)
+                    .end((err, res) => {
+                        if (err) {
+                            console.log(err);
+                            resolve(undefined);
+                        } else {
+                            resolve(JSON.parse(res.text));
+                        }
+                    })
+            );
+            assert(signInOptionsResponse.status === "OK");
+
+            const { createAndAssertCredential } = await getWebauthnLib();
+            const credential = createAndAssertCredential(registerOptionsResponse, signInOptionsResponse, {
```

Comment:
why does `createAndAssertCredential` take both the register and sign in options?
what exactly does it assert?

Reply:
It basically both creates  (used for signup)  and verifies a credential (used for login). 
I do it like this because the virtual authenticator doesn't store state and in order to verify (sign in) a credential, the virtual authenticator needs to have the credential stored (similar to how the phone stores the credential after creation). Doing it like this, i generate the credential and also verify using a single method, avoiding storing/handling state.

Reply:
I see, that makes it easier to work with wasm. One thing that still confuses me is that we re-use the same attestation for both sign up and in, but I'll try running the tests and check the actual object structure.